R/merge.R

```diff
@@ -147,39 +124,72 @@
     return(x)
 }
 
-# This function works similarly to scuttle::sumCountsAcrossFeatures but this
-# excludes NAs from the data. The scuttle function cannot handle NAs.
+.check_assay_for_merge <- function(assay.type, assay){
+    value.warning <- paste(c("\nAgglomeration of it might lead to meaningless ",
+        "values.\nCheck the assay, and consider doing transformation again ",
+        "manually with agglomerated data."))
+    # Check if assay includes binary values
+    if( all(assay == 0 | assay == 1) ){
+        warning("'", assay.type, "'", " includes binary values.",
+                value.warning, call. = FALSE)
+    }
+    # Check if assay includes negative values
+    if( !all(assay >= 0 | is.na(assay)) ){
+        warning("'", assay.type, "'", " includes negative values.",
+                value.warning, call. = FALSE)
+    }
+}
+
 #' @importFrom DelayedArray DelayedArray type rowsum
-.sum_counts_accross_features_na <- function(x, average, ids, ...){
-    # Which cell is not NA?
-    is_not_na <- !is.na(x)
-    type(is_not_na) <- "integer"
-    # Aggregate data to certain groups
-    x <- rowsum(x, ids, na.rm = TRUE)
-    # Calculate average if specified
-    if( average ){
-        x <- x/rowsum(is_not_na, ids)
+#' @importFrom scuttle sumCountsAcrossFeatures
+.agglomerate_assay <- function(
+    assay.type, assay, by, ids, na.rm, average, BPPARAM
+    ){
+    # Check assay
+    .check_assay_for_merge(assay.type, assay)
+    # Transpose if we are merging columns
+    if( by == 2L ){
+        assay <- t(assay)
     }
-    return(x)
```

Comment:
Is there a reason why the functionality from this function is moved? It is better to modify only necessary lines to improve clarity and maintainability while minimizing the risk of introducing unintended changes

Reply:
Sorry for the extra change. I was trying to make `.merge_rows_or_columns` more efficient, and give it a similar structure of `.agglomerate_module_assay` (which in my opinion is also clearer).

The code for `.merge_rows_or_columns` was using lapply for each of the 4 implemented stages of assay-wise operation (checking, transposing, merging and transposing back). For a clearer and more efficient syntax, I reduced the 4 lapply calls to a single streamlined (m)apply call, which in addition also checks and removes NAs for each assay (not only the first one).